File: 2018/08.scala
import scala.io.Source
import scala.collection.mutable.ArrayBuffer

object Solve06 {
  def main(args: Array[String]) {
    var file_input = Source.fromFile("08.in").mkString.split(" ").map(_.toInt)

    case class Node(children_count: Int, metadata_count: Int, children: ArrayBuffer[Node], metadata: Array[Int])


    def readNodes(input: Array[Int]): (Node, Array[Int]) = {
      var node = Node(input(0), input(1), ArrayBuffer.empty, input.takeRight(input(1)))
      for (x <- 0 to node.children_count) {
        var inner = input.drop(2).dropRight(node.metadata.length)
        println("reading inner length: " + inner.length)
        node.children += readNode(inner)
      }
      return node
    }

    var root = readNode(file_input)
    println(root.children_count)
    println(root.metadata_count)
    println(root.metadata)
  }
}

File: 2018/08.py
#!/usr/bin/env python3


def read_node(data):
    """Returns (node, unused_data)."""
    children = []
    metadata = []
    children_count = data[0]
    metadata_count = data[1]

    unused_data = data[2:]
    for _ in range(children_count):
        (node, unused_data) = read_node(unused_data)
        children.append(node)

    return (
        {"children": children, "metadata": unused_data[:metadata_count]},
        unused_data[metadata_count:])


def sum_metadata(node):
    return sum(node["metadata"]) + sum(map(sum_metadata, node["children"]))


def node_value(node):
    if not node["children"]:
        return sum(node["metadata"])
    s = 0
    for m in map(lambda x: x - 1, node["metadata"]):
        if m < len(node["children"]):
            s += node_value(node["children"][m])
    return s


with open("08.in") as f_in:
    data = map(int, f_in.read().strip().split(" "))

(root, unused) = read_node(data)

print(sum_metadata(root))
print(node_value(root))
